internal/api/grpc/server/gateway.go

```diff
@@ -149,3 +172,41 @@ func http1Host(next http.Handler, http1HostName string) http.Handler {
 		next.ServeHTTP(w, r)
 	})
 }
+
+func exhaustedCookieInterceptor(
+	next http.Handler,
+	cookieHandler *http_utils.CookieHandler,
+	cookieConfig *http_mw.AccessConfig,
+	queries *query.Queries,
+) http.Handler {
+	return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+		next.ServeHTTP(&cookieResponseWriter{
+			ResponseWriter: writer,
+			cookieHandler:  cookieHandler,
+			cookieConfig:   cookieConfig,
+			request:        request,
+			queries:        queries,
+		}, request)
+	})
+}
+
+type cookieResponseWriter struct {
+	http.ResponseWriter
+	cookieHandler *http_utils.CookieHandler
+	cookieConfig  *http_mw.AccessConfig
+	request       *http.Request
+	queries       *query.Queries
+}
+
+func (r *cookieResponseWriter) WriteHeader(status int) {
+	if status == 429 {
+		instance, err := r.queries.InstanceByHost(r.request.Context(), r.request.Host)
+		if err != nil {
+			logging.WithError(err).Warn("could not get instance for templating exhausted cookie value")
+		}
+		http_mw.SetExhaustedCookie(r.cookieHandler, r.ResponseWriter, r.cookieConfig, instance, r.request.Host)
+	} else {
+		http_mw.DeleteExhaustedCookie(r.cookieHandler, r.ResponseWriter, r.request, r.cookieConfig)
+	}
+	r.ResponseWriter.WriteHeader(status)
+}
```

Comment:
```suggestion
	if status != http.StatusTooManyRequests {
		http_mw.DeleteExhaustedCookie(r.cookieHandler, r.ResponseWriter, r.request, r.cookieConfig)
		r.ResponseWriter.WriteHeader(status)
		return
	}
	instance, err := r.queries.InstanceByHost(r.request.Context(), r.request.Host)
	logging.OnError(err).Warn("could not get instance for templating exhausted cookie value")
	http_mw.SetExhaustedCookie(r.cookieHandler, r.ResponseWriter, r.cookieConfig, instance, r.request.Host)
	r.ResponseWriter.WriteHeader(status)
}
```

Reply:
and just noticed, that the `host` for `r.queries.InstanceByHost(r.request.Context(), r.request.Host)` can actually be configured to be a custom header